### ISSUE TITLE
OpenMP bug in temperature calculations. 

### ISSUE THREAD

**Issue by pelahi** [ICRAR/VELOCIraptor-STF#117]
**Describe the bug**
There is an OpenMP bug with a variable not being declared thread private, see https://github.com/ICRAR/VELOCIraptor-STF/blob/44c4960f73393927ec76d7cfc0a66ae94f257560/src/substructureproperties.cxx#L1495-L1513
I have created a local branch with the fix but cannot generate a full PR as I am not a contributor to this repository. However, the change is a simple fix so I do not need to be granted contributor access to push a branch and generate a PR.

**Comment by pelahi** [comment]
@rtobar , did you want to add me as a contributor so I can generate a PR or are you happy to fix this?

**Comment by rtobar** [comment]
@pelahi thanks for reporting this!

Thanks for offering submitting a PR too. I'm half-assuming your change adds `temp` to the list of `private` group of the enclosing parallel omp pragma? My approach on these situations is different: I move the variables down to the innermost scope possible, inside the OpenMP parallel regions, thus avoiding any OpenMP-related shenanigans. This makes code easier to follow, and avoids OpenMP clutter around it, so I'd prefer if the fix goes in that direction. Something like this:

```diff
[rtobar@bolano VELOCIraptor-STF]---> git diff src/substructureproperties.cxx
diff --git a/src/substructureproperties.cxx b/src/substructureproperties.cxx
index 01b732f..b2aeae8 100644
--- a/src/substructureproperties.cxx
+++ b/src/substructureproperties.cxx
@@ -340,7 +340,7 @@ void GetProperties(Options &opt, const Int_t nbodies, Particle *Part, Int_t ngro
     Particle *Pval;
     Int_t i,j,k;
     Coordinate cmold(0.),cmref;
-    Double_t ri,rcmv,r2,cmx,cmy,cmz,EncMass,Ninside, SFR, temp;
+    Double_t ri,rcmv,r2,cmx,cmy,cmz,EncMass,Ninside, SFR;
     Double_t EncMassSF,EncMassNSF;
     Double_t cmvx,cmvy,cmvz;
     Double_t vc,rc,x,y,z,vx,vy,vz,jzval,Rdist,zdist,Ekin,Krot,mval;
@@ -413,7 +413,7 @@ private(i,j,k,Pval,cmref)
 #ifdef USEOPENMP
 #pragma omp parallel default(shared)  \
 private(i,j,k,Pval,ri,rcmv,r2,cmx,cmy,cmz,EncMass,Ninside,cmold,change,tol)\
-private(x,y,z,vx,vy,vz,vc,rc,jval,jzval,Rdist,zdist,Ekin,Krot,mval,RV_Ekin,RV_Krot,RV_num,SFR,temp)\
+private(x,y,z,vx,vy,vz,vc,rc,jval,jzval,Rdist,zdist,Ekin,Krot,mval,RV_Ekin,RV_Krot,RV_num,SFR)\
 private(EncMassSF,EncMassNSF,Krot_sf,Krot_nsf,Ekin_sf,Ekin_nsf)
 {
     #pragma omp for schedule(dynamic) nowait
@@ -628,7 +628,7 @@ private(EncMassSF,EncMassNSF,Krot_sf,Krot_nsf,Ekin_sf,Ekin_nsf)
                 mval = opt.MassValue;
 #endif
                 //store temperature
-                temp = Pval->GetTemperature();
+                auto temp = Pval->GetTemperature();
                 pdata[i].Temp_gas+=temp;
                 pdata[i].Temp_mean_gas+=mval*temp;
                 //pdata[i].sphden_gas+=Pval->GetMass()*Pval->GetSPHDen();
@@ -1473,7 +1473,7 @@ private(j,Pval,mval,x,y,z,vx,vy,vz,jval,jzval,zdist,Rdist)
                 else pdata[i].M_gas_nsf+=mval;
                 #endif
                 #if (defined(GASON)) || (defined(GASON) && defined(SWIFTINTERFACE))
-               temp=Pval->GetTemperature();
+               auto temp = Pval->GetTemperature();
                 if (temp > opt.temp_max_cut && SFR <= 0) {
                        pdata[i].M_gas_highT+=mval;
                }
@@ -1508,7 +1508,7 @@ private(j,Pval,x,y,z,vx,vy,vz,J,mval,SFR)
 #endif
 #ifdef STARON
                 SFR=Pval->GetSFR();
-               temp=Pval->GetTemperature();
+               auto temp = Pval->GetTemperature();
 #endif
 
                 x = (*Pval).X();
```

Please let me know if this fixes the issue. If so, and if you're OK with it, I'd rather go with this change, unless your local changes also follow this pattern. I might be missing something silly too, so please let me know if this looks OK.

**Comment by pelahi** [comment]
I actually was going to suggest this rather than adding temp to private but wasn't certain if you wanted to keep it the same as the previous lines at 628. So this should fix it. I am going to clean up this code as well just to make it more robust. Something like 
```c++
//vectors to store index of objects that should be processed by single thread or multiple threads
std::vector<Int_t> activeobjectlist, activeompobjectlist; 

#pragma omp parallel for \
default(none) \
shared(...) \ 
schedule(dynamic,1) \ 
if (nthreads > 1 && numgroups > omp_num_limit) 
#endif
for (auto i:activeobjectlist) {
   CalculatePropertiesSingleObject(...); 
}
``` 
And encapsulate all the thread private quantities within a single function for larger  objects will have something similar. I am even thinking of using a taskloop and just assign number of threads based on size of the object. 

Thoughts?

**Comment by rtobar** [comment]
Closed via #119

**Comment by rtobar** [comment]
> I actually was going to suggest this rather than adding temp to private but wasn't certain if you wanted to keep it the same as the previous lines at 628. So this should fix it. I am going to clean up this code as well just to make it more robust. Something like
> 
> ```c++
> //vectors to store index of objects that should be processed by single thread or multiple threads
> std::vector<Int_t> activeobjectlist, activeompobjectlist; 
> 
> #pragma omp parallel for \
> default(none) \
> shared(...) \ 
> schedule(dynamic,1) \ 
> if (nthreads > 1 && numgroups > omp_num_limit) 
> #endif
> for (auto i:activeobjectlist) {
>    CalculatePropertiesSingleObject(...); 
> }
> ```
> 
> And encapsulate all the thread private quantities within a single function for larger objects will have something similar. I am even thinking of using a taskloop and just assign number of threads based on size of the object.
> 
> Thoughts?

Having variables defined in the innermost scope possible is definitely better, I agree. If you want to factor out some code into a function that's fine too of course. I'm not sure I follow your statements about single/multiple threads and dealing with objects of different sizes, but I imagine this is all to do with balancing the load as much as possible. If you want to go there I'd suggest you clearly quantify things (balancing, speedup, etc.) before changing things, otherwise it's not clear if some changes yield improvements or not, or even if there's a benefit on spending time on trying to improve the code.
